patches/tModLoader/Terraria/Item.TML.cs

```diff
@@ -295,4 +297,29 @@ private void RestoreMeleeSpeedBehaviorOnVanillaItems()
 				attackSpeedOnlyAffectsWeaponAnimation = true;
 		}
 	}
+
+	// Added by TML.
+	/// <summary>
+	/// Determines the <see cref="PrefixCategory">prefix categories</see> of this <see cref="Item"/>.
+	/// </summary>
+	/// <returns>A <see cref="List{PrefixCategory}"/> of every category this <see cref="Item"/> matches, the <see cref="List{PrefixCategory}"/> will be empty if this <see cref="Item"/> doesn't have any categories.</returns>
+	public List<PrefixCategory> GetPrefixCategories() {
+		List<PrefixCategory> categories = new List<PrefixCategory>();
+		if (PrefixLegacy.ItemSets.SwordsHammersAxesPicks[type] || ItemLoader.MeleePrefix(this))
+			categories.Add(PrefixCategory.Melee);
+
+		if (PrefixLegacy.ItemSets.GunsBows[type] || ItemLoader.RangedPrefix(this))
+			categories.Add(PrefixCategory.Ranged);
+
+		if (PrefixLegacy.ItemSets.MagicAndSummon[type] || ItemLoader.MagicPrefix(this))
+			categories.Add(PrefixCategory.Magic);
+
+		if (PrefixLegacy.ItemSets.SpearsMacesChainsawsDrillsPunchCannon[type] || PrefixLegacy.ItemSets.BoomerangsChakrams[type] || PrefixLegacy.ItemSets.ItemsThatCanHaveLegendary2[type] || ItemLoader.WeaponPrefix(this) || categories.Count != 0)
+			categories.Add(PrefixCategory.AnyWeapon);
+
+		if (IsAPrefixableAccessory())
+			categories.Add(PrefixCategory.Accessory);
+
+		return categories;
+	}
```

Comment:
Can we bake this into a `static List<PrefixCategory>[]` or `PrefixCategory[][]` by type? Perhaps with some de-duplication?

Reply:
If I'm understanding you correctly and you mean an array mapping item IDs to the collections of prefix categories that item type can get, I should be able to do that pretty quickly;
If you haven't said that I've misunderstood by the next opportunity I have to implement this, it will be done shortly;
I didn't initially consider it because it seemed more practical to avoid spreading changes out across more files than necessary or making maintaining this part of maintaining something else, but I was forgetting that modders regularly find ways to make any performance cost problematic.

Reply:
Yes, I didn't identify any high traffic call sites in tML, but it seems like the thing we can do fairly easily to avoid allocs. Modders don't tend to be too perf aware. Filling out an array during `PostSetupContent` should suffice

Reply:
Updated to a 'compute if not baked' approach;
I'm not certain if this is the most performant possible implementation, but it should definitely be better for anything that spams GetPrefixCategories than the previous one